keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,517 @@
+
+## Motivation
+
+
+### Goals
+
+* Cloud-provider independent specification of basic QoS parameters
+* Specifying QoS at volume creation
+* Changing QoS parameters on-line for existing volumes
+* Specify quotas in a similar way to that of volume capacity
+
+### Non-Goals
+
+* OS-level IO QoS, for example cgroup policy
+* Inter-pod QoS. This KEP is restricted to what QoS can be defined on a volume.
+* Volume QoS classes. This proposal is for fine-grained provisioned IO control, and not for defining general classes of volume performance. Currently, storage vendors may accomplish this by adding custom parameters to a storage class. A future KEP may codify such behavior with common storage class parameter names, but that is out of scope of this KEP.
+* Scheduling based on total IO limits of a node. For example, total throughput for many cloud provider network attached storage volumes is limited by instance network bandwidth and may prevent achieving the provisioned IO. Scheduling to mitigate these factors is out of scope but it should be considered as a future extension of this KEP.
+
+## Proposal
+
+### Kubernetes API
+Add two well-known ResourceName keys to the existing ResourceList type used in PersistentVolumeClaim.Spec.Resources.{Limits,Requests} and PersistentVolume.Spec.Capacity:
+* ProvisionedIOPS
+* ProvisionedThroughput
+These values can be changed on a PersistentVolumeClaim, and will be reconciled against the storage provider in the same way as storage capacity resizing. The status of a PVC will reflect the currently provisioned IO as reported by the storage provider (in the same way that the status states the current size of the volume during a resize operation).
+
+```
+// Example:
+kind: PersistentVolumeClaim
+...
+spec:
+  ...
+  resources:
+requests:
+  storage: 4Gi
+  throughput: "1.2M"
+  iops: "160000"
+// Vendors may add additional resources using Device Plugin syntax: vendor-domain/resourcetype
+```
+
+[Device Plugin](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) is out of scope for this KEP and it is one of the options for configure vendor specific storage options.
+
+Add PVC Status field; change ResizeStatus alpha field to AllocatedResourceStatus map.
+
+### CSI API
+The CSI create request will be extended to add provisioned IO parameters. A new ControllerUpdateIOProvisioning RPC will be added.
+
+```
+// ControllerServer is the server API for Controller service.
+type ControllerServer interface {
+        ...
+        ControllerUpdateIOProvisioning(context.Context, *ControllerUpdateIOProvisioningRequest) (*ControllerUpdateIOProvisioningResponse, error)
```

Comment:
I wonder if we should add a ModifyVolume PRC but only allow IO changes for now.  In the future, if we decide to support other updates, we don't have to add another PRC.

Reply:
We can discuss this in today's meeting.